x-pack/filebeat/input/awss3/collector.go

```diff
@@ -93,7 +114,110 @@ type s3Context struct {
 	errC chan<- error
 }
 
-func (c *s3Collector) run() {
+type logEvent struct {
+	log    string
+	offset int64
+}
+
+func (c *s3BucketCollector) getS3Objects() ([]s3Info, error) {
+	var s3Infos []s3Info
+	bucketMetadata := strings.Split(c.config.S3Bucket, ":")
+	bucketName := bucketMetadata[len(bucketMetadata)-1]
+
+	listObjectInput := &s3.ListObjectsInput{
+		Bucket: awssdk.String(bucketName),
+	}
+
+	req := c.s3.ListObjectsRequest(listObjectInput)
+	paginator := s3.NewListObjectsPaginator(req)
+
+	for paginator.Next(c.cancellation) {
+		page := paginator.CurrentPage()
+		for _, object := range page.Contents {
+			// Unescape substrings from s3 log name. For example, convert "%3D" back to "="
+			filename := url.PathEscape(*object.Key)
+
+			state := NewState(bucketName, *object.Key, *object.Size, *object.LastModified)
+			if !c.states.IsNew(state) {
+				continue
+			}
+
+			c.states.Update(state)
+
+			info := s3Info{
+				name:         bucketName,
+				key:          filename,
+				arn:          c.config.S3Bucket,
+				readerConfig: c.config.ReaderConfig,
+			}
+
+			if len(c.config.FileSelectors) == 0 {
+				s3Infos = append(s3Infos, info)
+				continue
+			}
+
+			for _, fs := range c.config.FileSelectors {
+				if fs.Regex != nil && fs.Regex.MatchString(filename) {
+					s3Infos = append(s3Infos, info)
+
+					break
+				}
+			}
+		}
+	}
+
+	if err := paginator.Err(); err != nil {
+		return nil, errors.Wrap(err, "error FilterLogEvents with Paginator")
+	}
+
+	return s3Infos, nil
+}
+
+func (c *s3BucketCollector) run() {
+	defer c.logger.Info("s3 input worker has stopped.")
```

Comment:
should we also defer here in the info log between s3bucket collector vs s3sqs collector (line 221)?

Reply:
```
2021-07-27T10:28:48.376+0200    INFO    [input.aws-s3]  awss3/collector.go:201  s3 input worker has stopped.    {"id": "4F2455F48AA183A4", "s3_bucket": "arn:aws:s3:::mybucket"}
```
vs
```
2021-07-27T10:28:48.376+0200    INFO    [input.aws-s3]  awss3/collector.go:242  s3 input worker has stopped.    {"id": "4F2455F48AA183A4", "queue_url": "https://sqs.myregion.amazonaws.com/123456/myqueue"}

```

but I can add extra information about the collector type in the message